gnd/src/main/java/com/google/android/gnd/ui/home/featuresheet/RecordListViewModel.java

```diff
@@ -57,23 +71,48 @@ public void loadRecordSummaries(Feature feature, Form form) {
         feature.getProject(), feature.getFeatureType().getId(), form.getId(), feature.getId());
   }
 
+  /**
+   * Attempts to fetch a list of records based on the contents of a {@link RecordSummaryRequest}.
+   *
+   * @param request A record summary request. A triple of project, featureId, and formId.
+   * @return A list containing fetched records with forms that satisfy the formId
+   *     provided in the request.
+   */
+  private Single<List<Record>> fetchRecordSummaries(RecordSummaryRequest request) {
+    // TODO: Only fetch records with current formId.
+    return dataRepository
+        .getRecordSummaries(request.project.getId(), request.featureId)
+        .map(
+            records ->
+                stream(records)
+                    .filter(record -> record.getForm().getId().equals(request.formId))
+                    .collect(Collectors.toList()));
+  }
```

Comment:
While we're here can we extract this lambda into a helper for better readability and less nesting?

Reply:
Gotcha. I also discovered there's a Bifunction version of Flatmap that succinctly expresses what we're doing here: using the initial request emissions to produce a record list, then using some other component of the initial emissions (requests) to make a 'selection' over the record lists (in this case filtering by the form ID contained in the request).

This eliminates the need for nesting to capture previous scope and allows us to use method references instead of lambdas and function calls.